forcetree.c

```diff
@@ -144,75 +156,155 @@ int force_tree_build(int npart)
+        nfreep->u.suns[j] = -1;
 }
 
 /* Size of the free Node thread cache.
  * 100 was found to be optimal for an Intel skylake with 4 threads.*/
 #define NODECACHE_SIZE 100
 
-/*! Constructs the gravitational oct-tree.
- *
- *  The index convention for accessing tree nodes is the following: the
- *  indices 0...NumPart-1 reference single particles, the indices
- *  All.MaxPart.... All.MaxPart+nodes-1 reference tree nodes. `Nodes_base'
- *  points to the first tree node, while `nodes' is shifted such that
- *  nodes[All.MaxPart] gives the first tree node. Finally, node indices
- *  with values 'All.MaxPart + MaxNodes' and larger indicate "pseudo
- *  particles", i.e. multipole moments of top-level nodes that lie on
- *  different CPUs. If such a node needs to be opened, the corresponding
- *  particle must be exported to that CPU. */
-int force_tree_build_single(int npart)
+/*Get a pointer to memory for a free node, from our node cache.
+ * If there is no memory left, return NULL.*/
+int get_freenode(int * nnext, int *nnext_thread, int *nrem_thread)
+{
+    /*Get memory for an extra node from our cache.*/
+    if(*nrem_thread == 0) {
+        *nnext_thread = atomic_fetch_and_add(nnext, NODECACHE_SIZE);
+        *nrem_thread = NODECACHE_SIZE;
+    }
+    const int ninsert = (*nnext_thread)++;
+    (*nrem_thread)--;
+    return ninsert;
+}
+
+/* Parent is a node where the subnode we want to add a particle to is filled.
+ * We add a new internal node at this subnode and try to add both the old and new particles to it.
+ * Parent is assumed to be locked.*/
+int insert_internal_node(int parent, int subnode, int p_child, int p_toplace, const int lastnode, int *nnext, int *nnext_thread, int *nrem_thread, double minlen)
+{
+    /*Get memory for an extra node from our cache.*/
+    int ninsert = get_freenode(nnext, nnext_thread, nrem_thread);
+    /*If we already have too many nodes, exit loop.*/
+    if(*nnext_thread >= lastnode)
+        return 1;
+
+    struct NODE *nfreep = &Nodes[ninsert];
+    struct NODE *nprnt = &Nodes[parent];
+    /* We create a new internal node with empty subnodes at the end of the array, and
+     * use it to replace the particle in the parent's subnode.*/
+    init_internal_node(nfreep, nprnt, subnode);
+    /*Set father of new node*/
+    nfreep->father = parent;
+
+    /* The new internal node replaced a particle in the parent.
+     * Re-add that particle to the child.*/
+    const int child_subnode = get_subnode(nfreep, p_child);
+
+    int new_subnode = get_subnode(nfreep, p_toplace);
+
+    /* If the node is very small, just add the particle to the next subnode.
+     * This ensures that we can always construct the tree, and only happens far
+     * below the softening length. This happens occasionally, but if it is
+     * frequent increase the force softening.
+     */
+    if(nfreep->len < minlen && new_subnode == child_subnode) {
+        if(child_subnode < 7)
+            new_subnode = child_subnode + 1;
```

Comment:
I feel the bucketing shall be added before a node is split. Currently we will create a lot of empty nodes when many particles are within the ambiguity radius.

Reply:
If we do it before the node is split, we cannot guarantee that there is an empty subnode to receive the new particle. And if we create a new node, where do we do it? If there are enough particles within the ambiguity radius to really inflate the tree length, we have Troubles.

Reply:
We can. -- keep going till we find first empty slot but doesn't require it to be a correct child. e.g., adding a fucntion `get_any_empty_or_subnode()`